backends/ebpf/psa/externs/ebpfPsaHashAlgorithm.cpp

```diff
@@ -0,0 +1,251 @@
+    j = 1;
+
+    for (; i != 0; i >>=1, j <<= 1) {
+        if (i & poly)
+            result |= j;
+    }
+
+    return Util::printf_format("%llu", result);
+}
+
+void CRCChecksumAlgorithm::emitUpdateMethod(CodeBuilder* builder, int crcWidth) {
+    // Note that this update method is optimized for our CRC16 and CRC32, custom
+    // version may require other method of update. To deal with byte order data
+    // is read from the end of buffer.
+    cstring code = "static __always_inline\n"
+                   "void crc%w%_update(u%w% * reg, const u8 * data, "
+                   "u16 data_size, const u%w% poly) {\n"
+                   "    data += data_size - 1;\n"
+                   "    #pragma clang loop unroll(full)\n"
+                   "    for (u16 i = 0; i < data_size; i++) {\n"
+                   "        bpf_trace_message(\"CRC%w%: data byte: %x\\n\", *data);\n"
+                   "        *reg ^= *data;\n"
+                   "        for (u8 bit = 0; bit < 8; bit++) {\n"
+                   "            *reg = (*reg) & 1 ? ((*reg) >> 1) ^ poly : (*reg) >> 1;\n"
+                   "        }\n"
+                   "        data--;\n"
+                   "    }\n"
+                   "}";
+    code = code.replace("%w%", Util::printf_format("%d", crcWidth));
+    builder->appendLine(code);
+}
+
+void CRCChecksumAlgorithm::emitVariables(CodeBuilder* builder,
+                                         const IR::Declaration_Instance* decl) {
+    registerVar = program->refMap->newName(baseName + "_reg");
+
+    builder->emitIndent();
+
+    if (decl != nullptr) {
+        BUG_CHECK(decl->type->is<IR::Type_Specialized>(), "Must be a specialized type %1%", decl);
+        auto ts = decl->type->to<IR::Type_Specialized>();
+        BUG_CHECK(ts->arguments->size() == 1, "Expected 1 specialized type %1%", decl);
+
+        auto otype = ts->arguments->at(0);
+        if (!otype->is<IR::Type_Bits>()) {
+            ::error(ErrorType::ERR_UNSUPPORTED, "Must be bit or int type: %1%", ts);
+            return;
+        }
+        if (otype->width_bits() != crcWidth) {
+            ::error(ErrorType::ERR_TYPE_ERROR, "Must be %1%-bits width: %2%", crcWidth, ts);
+            return;
+        }
+
+        auto registerType = EBPFTypeFactory::instance->create(otype);
+        registerType->emit(builder);
+    } else {
+        if (crcWidth == 16)
+            builder->append("u16");
+        else if (crcWidth == 32)
+            builder->append("u32");
+        else
+            BUG("Unsupported CRC width %1%", crcWidth);
+    }
+
+    builder->appendFormat(" %s = %s", registerVar.c_str(), initialValue.c_str());
+    builder->endOfStatement(true);
+}
+
+void CRCChecksumAlgorithm::emitClear(CodeBuilder* builder) {
+    builder->emitIndent();
+    builder->appendFormat("%s = %s", registerVar.c_str(), initialValue.c_str());
+    builder->endOfStatement(true);
+}
+
+void CRCChecksumAlgorithm::emitAddData(CodeBuilder* builder,
```

Comment:
This code has several layers of indirection for generating a string that becomes C code, so it's not easy to read.
Can you add a comment showing some final generated code?